### DISCUSSION TITLE
Why Node.js ?

### DISCUSSION THREAD

**Discussion by mgcdanny** [uwdata/mosaic#150] [Q&A] [answered]
I am just wondering why the decision was made to wrap DuckDB around a node.js server.

If duckdb is a cpu process, wouldn't that block the Node.js event loop?

**Comment by domoritz** [comment, marked as answer]
We chose Node for the server because the part that runs in the browser is all JS already so we didn't have to use multiple toolchains. 

Having said that, the mosaic server is just a very thin wrapper around DuckDB. In our experiments it doesn't seem to be an issue that the DB is running in the same process. DuckDB uses multiple threats so it doesn't seem to block. I implemented a similar wrapper in Python for the widget, and you can find the code at https://github.com/uwdata/mosaic/blob/1590456b08bcb0bf4dd28ab8ac909b05a9ec0a69/packages/widget/mosaic_widget/__init__.py#L66. I can imagine that we implement a server in e.g. Rust in the future.
